# Finding All Relevant Data Sets
___

Data on the SAT scores and on the demographics for each high school are available online (6th and 3rd in the table below). However, both data sets alone are not sufficient for this project. Hence, we will need supplementary data to conduct our analysis. 

The table below contains the filenames of the datasets we will be using. This is followed by brief descriptions of and links to each dataset. 

|File Name|Description|Link to Dataset|
|:--------|:----------|:--------------|
|1. `ap_2010.csv`|\*Advanced Placement (AP) exam results for each high school|[Find Here](https://data.cityofnewyork.us/Education/2010-AP-College-Board-School-Level-Results/itfs-ms3e)
|2. `class_size.csv`|Information on class size  for each school|[Find Here](https://data.cityofnewyork.us/Education/2010-2011-Class-Size-School-level-detail/urz7-pzb3)
|3. `demographics.csv`|Demographic information for each school|[Find Here](https://data.cityofnewyork.us/Education/2006-2012-School-Demographics-and-Accountability-S/ihfw-zy9j)
|4. `graduation.csv`|The percentage of students who graduated, and other outcome information|[Find Here](https://data.cityofnewyork.us/Education/2005-2010-Graduation-Outcomes-School-Level/vh2h-md7a)|
|5. `hs_directory.csv`|A directory of high schools|[Find Here](https://data.cityofnewyork.us/Education/2014-2015-DOE-High-School-Directory/n3p6-zve2) 
|6. `sat_results.csv`|SAT scores for each high school in New York City|[Find Here](https://data.cityofnewyork.us/Education/2012-SAT-Results/f9bf-2cp4) 
|7. `survey_all.txt`|Surveys of parents, teachers, and students at each school|[Find Here](https://data.cityofnewyork.us/Education/2011-NYC-School-Survey/mnz3-dyi8)
|8. `school_districts.geojson`|Contains geographical data of NYC school districts for mapping|[Find Here](https://data.cityofnewyork.us/Education/School-Districts/r8nu-ymqj)


\*passing an optional AP exam in a particular subject can earn a student college credit in that subject

To facilitate our analysis and making finding correlations easier, we will combine all the data sets into a single one. Before doing that, we will have to clean and pre-process each data set. Note that our analysis will be high-level which means that we will not be particularly interested in granular data. This will be reflected in how we process the data sets below and how we will lose some data when combining the data sets 

# Reading in and Exploring the CSV Files
___

The table above shows that there are two file types: `csv` and `txt`. We will read in the `csv` files first. 

In [1]:
import pandas as pd
import numpy as np

data_files = [
    "ap_2010.csv",
    "class_size.csv",
    "demographics.csv",
    "graduation.csv",
    "hs_directory.csv",
    "sat_results.csv"
]

# Store the datasets as dataframes in a dictionary for convenience 
data = {} 
for f in data_files:
    d = pd.read_csv("datasets/{0}".format(f)) # The files are located in a folder named "datasets"
    data[f.replace(".csv", "")] = d

The first few rows of the SAT scores dataset (accessed with the dictionary) are displayed below to help us understand its structure. We will use the `display()` function instead of the `print()` function to make the results easier to read. 

In [2]:
display(data['sat_results'].head(10))

DBN                                        SCHOOL NAME  \
0  01M292      HENRY STREET SCHOOL FOR INTERNATIONAL STUDIES   
1  01M448                UNIVERSITY NEIGHBORHOOD HIGH SCHOOL   
2  01M450                         EAST SIDE COMMUNITY SCHOOL   
3  01M458                          FORSYTH SATELLITE ACADEMY   
4  01M509                            MARTA VALLE HIGH SCHOOL   
5  01M515            LOWER EAST SIDE PREPARATORY HIGH SCHOOL   
6  01M539  NEW EXPLORATIONS INTO SCIENCE, TECHNOLOGY AND ...   
7  01M650                               CASCADES HIGH SCHOOL   
8  01M696                     BARD HIGH SCHOOL EARLY COLLEGE   
9  02M047  47 THE AMERICAN SIGN LANGUAGE AND ENGLISH SECO...   

  Num of SAT Test Takers SAT Critical Reading Avg. Score SAT Math Avg. Score  \
0                     29                             355                 404   
1                     91                             383                 423   
2                     70                             377                 402   
3                      7                             414                 401   
4                     44                             390                 433   
5                    112                             332                 557   
6                    159                             522                 574   
7                     18                             417                 418   
8                    130                             624                 604   
9                     16                             395                 400   

  SAT Writing Avg. Score  
0                    363  
1                    366  
2                    370  
3                    359  
4                    384  
5                    316  
6                    525  
7                    411  
8                    628  
9                    387

Based on the results above, we can make the following observations:

- The school names tell us that only high schools are present in this data set. 
- Each school seems to have its own unique DBN value
- Each school appears only once which means that the same holds true for the DBN values. 
- For easier analysis, the last three columns containing the SAT scores can be combined into a single one. **(Each section has a total of 800 points, making the combined SAT score have a highest possible value of 2400)**

Let's explore the rest of the datasets below. 

In [3]:
pd.options.display.max_columns = 100 # To avoid truncated output
for d in data:
    print('\n {0}'.format(d))
    display(data[d].head())


 ap_2010


DBN                             SchoolName AP Test Takers   \
0  01M448           UNIVERSITY NEIGHBORHOOD H.S.              39   
1  01M450                 EAST SIDE COMMUNITY HS              19   
2  01M515                    LOWER EASTSIDE PREP              24   
3  01M539         NEW EXPLORATIONS SCI,TECH,MATH             255   
4  02M296  High School of Hospitality Management               s   

  Total Exams Taken Number of Exams with scores 3 4 or 5  
0                49                                   10  
1                21                                    s  
2                26                                   24  
3               377                                  191  
4                 s                                    s


 class_size


CSD BOROUGH SCHOOL CODE                SCHOOL NAME GRADE  PROGRAM TYPE  \
0    1       M        M015  P.S. 015 Roberto Clemente     0K       GEN ED   
1    1       M        M015  P.S. 015 Roberto Clemente     0K          CTT   
2    1       M        M015  P.S. 015 Roberto Clemente     01       GEN ED   
3    1       M        M015  P.S. 015 Roberto Clemente     01          CTT   
4    1       M        M015  P.S. 015 Roberto Clemente     02       GEN ED   

  CORE SUBJECT (MS CORE and 9-12 ONLY) CORE COURSE (MS CORE and 9-12 ONLY)  \
0                                    -                                   -   
1                                    -                                   -   
2                                    -                                   -   
3                                    -                                   -   
4                                    -                                   -   

  SERVICE CATEGORY(K-9* ONLY)  NUMBER OF STUDENTS / SEATS FILLED  \
0                           -                               19.0   
1                           -                               21.0   
2                           -                               17.0   
3                           -                               17.0   
4                           -                               15.0   

   NUMBER OF SECTIONS  AVERAGE CLASS SIZE  SIZE OF SMALLEST CLASS  \
0                 1.0                19.0                    19.0   
1                 1.0                21.0                    21.0   
2                 1.0                17.0                    17.0   
3                 1.0                17.0                    17.0   
4                 1.0                15.0                    15.0   

   SIZE OF LARGEST CLASS DATA SOURCE  SCHOOLWIDE PUPIL-TEACHER RATIO  
0                   19.0         ATS                             NaN  
1                   21.0         ATS                             NaN  
2                   17.0         ATS                             NaN  
3                   17.0         ATS                             NaN  
4                   15.0         ATS                             NaN


 demographics


DBN                       Name  schoolyear fl_percent  frl_percent  \
0  01M015  P.S. 015 ROBERTO CLEMENTE    20052006       89.4          NaN   
1  01M015  P.S. 015 ROBERTO CLEMENTE    20062007       89.4          NaN   
2  01M015  P.S. 015 ROBERTO CLEMENTE    20072008       89.4          NaN   
3  01M015  P.S. 015 ROBERTO CLEMENTE    20082009       89.4          NaN   
4  01M015  P.S. 015 ROBERTO CLEMENTE    20092010                    96.5   

   total_enrollment prek   k grade1 grade2 grade3 grade4 grade5 grade6 grade7  \
0               281   15  36     40     33     38     52     29     38    NaN   
1               243   15  29     39     38     34     42     46    NaN    NaN   
2               261   18  43     39     36     38     47     40    NaN    NaN   
3               252   17  37     44     32     34     39     49    NaN    NaN   
4               208   16  40     28     32     30     24     38    NaN    NaN   

  grade8 grade9 grade10 grade11 grade12  ell_num  ell_percent  sped_num  \
0    NaN    NaN     NaN     NaN     NaN     36.0         12.8      57.0   
1    NaN    NaN     NaN     NaN     NaN     38.0         15.6      55.0   
2    NaN    NaN     NaN     NaN     NaN     52.0         19.9      60.0   
3    NaN    NaN     NaN     NaN     NaN     48.0         19.0      62.0   
4    NaN    NaN     NaN     NaN     NaN     40.0         19.2      46.0   

   sped_percent ctt_num selfcontained_num  asian_num  asian_per  black_num  \
0          20.3      25                 9         10        3.6         74   
1          22.6      19                15         18        7.4         68   
2          23.0      20                14         16        6.1         77   
3          24.6      21                17         16        6.3         75   
4          22.1      14                14         16        7.7         67   

   black_per  hispanic_num  hispanic_per  white_num  white_per  male_num  \
0       26.3           189          67.3          5        1.8     158.0   
1       28.0           153          63.0          4        1.6     140.0   
2       29.5           157          60.2          7        2.7     143.0   
3       29.8           149          59.1          7        2.8     149.0   
4       32.2           118          56.7          6        2.9     124.0   

   male_per  female_num  female_per  
0      56.2       123.0        43.8  
1      57.6       103.0        42.4  
2      54.8       118.0        45.2  
3      59.1       103.0        40.9  
4      59.6        84.0        40.4


 graduation


Demographic     DBN                            School Name    Cohort  \
0  Total Cohort  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL      2003   
1  Total Cohort  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL      2004   
2  Total Cohort  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL      2005   
3  Total Cohort  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL      2006   
4  Total Cohort  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL  2006 Aug   

   Total Cohort Total Grads - n Total Grads - % of cohort Total Regents - n  \
0             5               s                         s                 s   
1            55              37                     67.3%                17   
2            64              43                     67.2%                27   
3            78              43                     55.1%                36   
4            78              44                     56.4%                37   

  Total Regents - % of cohort Total Regents - % of grads Advanced Regents - n  \
0                           s                          s                    s   
1                       30.9%                      45.9%                    0   
2                       42.2%                      62.8%                    0   
3                       46.2%                      83.7%                    0   
4                       47.4%                      84.1%                    0   

  Advanced Regents - % of cohort Advanced Regents - % of grads  \
0                              s                             s   
1                             0%                            0%   
2                             0%                            0%   
3                             0%                            0%   
4                             0%                            0%   

  Regents w/o Advanced - n Regents w/o Advanced - % of cohort  \
0                        s                                  s   
1                       17                              30.9%   
2                       27                              42.2%   
3                       36                              46.2%   
4                       37                              47.4%   

  Regents w/o Advanced - % of grads Local - n Local - % of cohort  \
0                                 s         s                   s   
1                             45.9%        20               36.4%   
2                             62.8%        16                 25%   
3                             83.7%         7                  9%   
4                             84.1%         7                  9%   

    Local - % of grads Still Enrolled - n Still Enrolled - % of cohort  \
0                    s                  s                            s   
1                54.1%                 15                        27.3%   
2  37.200000000000003%                  9                        14.1%   
3                16.3%                 16                        20.5%   
4                15.9%                 15                        19.2%   

  Dropped Out - n Dropped Out - % of cohort  
0               s                         s  
1               3                      5.5%  
2               9                     14.1%  
3              11                     14.1%  
4              11                     14.1%


 hs_directory


dbn                                        school_name       boro  \
0  17K548                Brooklyn School for Music & Theatre   Brooklyn   
1  09X543                   High School for Violin and Dance      Bronx   
2  09X327        Comprehensive Model School Project M.S. 327      Bronx   
3  02M280     Manhattan Early College School for Advertising  Manhattan   
4  28Q680  Queens Gateway to Health Sciences Secondary Sc...     Queens   

  building_code    phone_number    fax_number grade_span_min  grade_span_max  \
0          K440    718-230-6250  718-230-6262              9              12   
1          X400    718-842-0687  718-589-9849              9              12   
2          X240    718-294-8111  718-294-8109              6              12   
3          M520  718-935-3477             NaN              9              10   
4          Q695    718-969-3155  718-969-3552              6              12   

  expgrade_span_min  expgrade_span_max  \
0               NaN                NaN   
1               NaN                NaN   
2               NaN                NaN   
3                 9               14.0   
4               NaN                NaN   

                                                 bus  \
0              B41, B43, B44-SBS, B45, B48, B49, B69   
1  Bx13, Bx15, Bx17, Bx21, Bx35, Bx4, Bx41, Bx4A,...   
2  Bx1, Bx11, Bx13, Bx18, Bx2, Bx3, Bx32, Bx35, Bx36   
3                             M103, M15, M22, M5, M9   
4                                      Q25, Q46, Q65   

                                              subway  primary_address_line_1  \
0  2, 3, 4, 5, F, S to Botanic Garden ; B, Q to P...      883 Classon Avenue   
1                              2, 5 to Intervale Ave        1110 Boston Road   
2                4 to Mt Eden Ave ; B, D to 170th St      1501 Jerome Avenue   
3  1 to Chambers St ; 2, 3, A, C to Fulton St-Bro...        411 Pearl Street   
4                                                NaN  160-20 Goethals Avenue   

       city state_code    zip                                       website  \
0  Brooklyn         NY  11225                           Bkmusicntheatre.com   
1     Bronx         NY  10456                                  www.hsvd.org   
2     Bronx         NY  10452  http://schools.nyc.gov/schoolportals/09/X327   
3  New York         NY  10038                                           NaN   
4   Jamaica         NY  11432                         www.queensgateway.com   

   total_students                          campus_name  \
0           399.0  Prospect Heights Educational Campus   
1           378.0            Morris Educational Campus   
2           543.0  DOE New Settlement Community Campus   
3             NaN   Murry Bergtraum Educational Campus   
4           811.0                                  NaN   

                   school_type  \
0                          NaN   
1                          NaN   
2                          NaN   
3  NYC P-Tech 9-14, New School   
4                          NaN   

                                  overview_paragraph  \
0  Brooklyn School for Music & Theatre (BSMT) use...   
1  The High School for Violin and Dance (HSVD) is...   
2  At the Comprehensive Model School Project (CMS...   
3  Through close partnerships with the Borough of...   
4  Queens Gateway to Health Sciences Secondary Sc...   

                                  program_highlights  \
0  We offer highly competitive positions in our D...   
1  Freshmen take both violin and dance; College N...   
2  After-school and Saturday Tutoring, Advisory, ...   
3  Early College experience leading to a free ass...   
4  PSAT/SAT Prep (grades 9-11 must take the PSAT ...   

                                    language_classes  \
0                                            Spanish   
1                                            Spanish   
2                                            Spanish   
3  N/A (Expected: Chinese [Mandarin], French, Spa...   
4                   


 sat_results


DBN                                    SCHOOL NAME  \
0  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL STUDIES   
1  01M448            UNIVERSITY NEIGHBORHOOD HIGH SCHOOL   
2  01M450                     EAST SIDE COMMUNITY SCHOOL   
3  01M458                      FORSYTH SATELLITE ACADEMY   
4  01M509                        MARTA VALLE HIGH SCHOOL   

  Num of SAT Test Takers SAT Critical Reading Avg. Score SAT Math Avg. Score  \
0                     29                             355                 404   
1                     91                             383                 423   
2                     70                             377                 402   
3                      7                             414                 401   
4                     44                             390                 433   

  SAT Writing Avg. Score  
0                    363  
1                    366  
2                    370  
3                    359  
4                    384

We can make some observations based on the first few rows of each one.

- Except for `class_size`, all the data sets contain a DBN column. However, `class_size` has columns named `CSD` and `SCHOOL CODE` that can be used to create a DBN. Once all data sets have a DBN column, this can be used as a reference column to combine them. 
- Some data sets, such as `graduation`, have high schools that are listed multiple times. We will have to condense these data sets so that the unique DBN values only appear once per data set. Failing to do this will result in an error when combining the data sets. 
- `hs_directory` has a column at the very end, `Location 1`, that contains geographical coordinates. These can be used for mapping. 

# Reading In and Exploring the Survey
___

Next, we will read in and explore the `txt` file, or the survey file. Since this file is tab delimited and `windows-1252` encoded, these need to be specified when reading in. 

In [4]:
# Read in the survey
survey_all = pd.read_csv("datasets/survey_all.txt", delimiter="\t", encoding='windows-1252')

display(survey_all.head())
survey_all.shape

dbn    bn                      schoolname  d75 studentssurveyed  \
0  01M015  M015       P.S. 015 Roberto Clemente    0               No   
1  01M019  M019             P.S. 019 Asher Levy    0               No   
2  01M020  M020            P.S. 020 Anna Silver    0               No   
3  01M034  M034  P.S. 034 Franklin D. Roosevelt    0              Yes   
4  01M063  M063       P.S. 063 William McKinley    0               No   

   highschool                  schooltype  rr_s  rr_t  rr_p    N_s   N_t  \
0         0.0           Elementary School   NaN    88    60    NaN  22.0   
1         0.0           Elementary School   NaN   100    60    NaN  34.0   
2         0.0           Elementary School   NaN    88    73    NaN  42.0   
3         0.0  Elementary / Middle School  89.0    73    50  145.0  29.0   
4         0.0           Elementary School   NaN   100    60    NaN  23.0   

     N_p  nr_s  nr_t  nr_p  saf_p_11  com_p_11  eng_p_11  aca_p_11  saf_t_11  \
0   90.0     0    25   150       8.5       7.6       7.5       7.8       7.5   
1  161.0     0    33   269       8.4       7.6       7.6       7.8       8.6   
2  367.0     0    48   505       8.9       8.3       8.3       8.6       7.6   
3  151.0   163    40   301       8.8       8.2       8.0       8.5       7.0   
4   90.0     0    23   151       8.7       7.9       8.1       7.9       8.4   

   com_t_11  eng_t_11  aca_t_11  saf_s_11  com_s_11  eng_s_11  aca_s_11  \
0       7.8       7.6       7.9       NaN       NaN       NaN       NaN   
1       8.5       8.9       9.1       NaN       NaN       NaN       NaN   
2       6.3       6.8       7.5       NaN       NaN       NaN       NaN   
3       6.2       6.8       7.8       6.2       5.9       6.5       7.4   
4       7.3       7.8       8.1       NaN       NaN       NaN       NaN   

   saf_tot_11  com_tot_11  eng_tot_11  aca_tot_11  p_q2h  p_q7a  p_q7b  p_q7c  \
0         8.0         7.7         7.5         7.9    8.0    8.2    8.3    7.5   
1         8.5         8.1         8.2         8.4    7.7    7.9    8.0    7.3   
2         8.2         7.3         7.5         8.0    8.1    8.8    8.9    8.5   
3         7.3         6.7         7.1         7.9    8.1    8.5    8.8    8.2   
4         8.5         7.6         7.9         8.0    8.0    8.4    8.6    8.0   

   p_q7d  p_q8a  p_q8b  p_q8c  p_q8d  p_q8e  p_q8f  p_q2b  p_q2d  p_q2e  \
0    7.9    6.8    8.7    9.7    8.7    9.9    9.9    7.7    8.3    7.9   
1    7.7    6.5    8.8    9.4    8.7   10.0    9.9    7.5    8.2    7.9   
2    8.4    7.6    9.2    9.4    9.2    9.8    9.7    8.4    8.8    8.5   
3    8.3    7.3    9.2    9.4    9.1    9.8    9.7    8.3    8.7    8.3   
4    8.0    6.5    8.8    9.6    9.4   10.0   10.0    7.6    8.5    8.0   

   p_q2f  p_q2g  p_q3a  p_q3b  ...  s_N_q13b_3  s_N_q13b_4  s_N_q13c_1  \
0    8.1    7.5    7.3    6.7  ...         NaN         NaN         NaN   
1    8.0    7.2    7.0    6.9  ...         NaN         NaN         NaN   
2    8.3    8.0    7.6    7.3  ...         NaN         NaN         NaN   
3    8.1    7.6    7.6    7.5  ...        19.0        16.0        24.0   
4    8.2    7.4    7.8    7.2  ...         NaN         NaN         NaN   

   s_N_q13c_2  s_N_q13c_3  s_N_q13c_4  s_N_q13d_1  s_N_q13d_2  s_N_q13d_3  \
0         NaN         NaN         NaN         NaN         NaN         NaN   
1         NaN         NaN         NaN         NaN         NaN         NaN   
2         NaN         NaN         NaN         NaN         NaN         NaN   
3        74.0        20.0        24.0        21.0        50.0        26.0   
4         NaN         NaN         NaN         NaN         NaN         NaN   

   s_N_q13d_4  s_N_q13e_1  s_N_q13e_2  s_N_q13e_3  s_N_q13e_4  s_N_q13f_1  \
0         NaN         NaN         NaN         NaN         NaN         NaN   
1         NaN         NaN         NaN         NaN         NaN         NaN   
2         NaN         NaN         NaN         NaN         NaN         NaN   
3        45.0        76.

(1646, 1942)

Two observations can immediately be made:

- There are almost 2000 columns, most of which we do not need. We will want retain only those needed for our analysis. Printing the data set will also be easier with fewer columns.  
- We will want to convert the column name, `dbn`, to an uppercase `DBN` to be consistent with the other data sets. 

To aid us in selecting the columns to be retained, we can use the [data dictionary](https://data.cityofnewyork.us/api/views/mnz3-dyi8/files/aa68d821-4dbb-4eb2-9448-3d8cbbad5044?download=true&filename=Survey%20Data%20Dictionary.xls) that was uploaded with the [survey for all districts](https://data.cityofnewyork.us/Education/2011-NYC-School-Survey/mnz3-dyi8). Using the objective of this project as our basis, the names of the selected columns are in the list `survey_fields` in the code cell below. These columns will give us aggregate survey data about how parents, teachers, and students feel about school safety, academic performance, and more. The `DBN` column is also included.

In [5]:
# Rename 'dbn' to uppercase 'DBN'
survey_all.rename(columns = {'dbn':'DBN'}, inplace = True)

# List of columns to be used for filtering
survey_fields = [
    "DBN", 
    "rr_s", 
    "rr_t", 
    "rr_p", 
    "N_s", 
    "N_t", 
    "N_p", 
    "saf_p_11", 
    "com_p_11", 
    "eng_p_11", 
    "aca_p_11", 
    "saf_t_11", 
    "com_t_11", 
    "eng_t_11", 
    "aca_t_11", 
    "saf_s_11", 
    "com_s_11", 
    "eng_s_11", 
    "aca_s_11", 
    "saf_tot_11", 
    "com_tot_11", 
    "eng_tot_11", 
    "aca_tot_11",
]

# Filter the combined the dataset and reassign it to the data dictionary
survey = survey_all.loc[:,survey_fields]
data['survey'] = survey

data['survey'].shape

(1646, 23)

The number of columns has now been reduced to 23 from the previous ~2000. 

# Add DBN columns
___

When we explored all of the data sets, we noticed that `class_size` does not have a `DBN` column. `hs_directory` has a `dbn` column that just needs to be renamed to an uppercase format.

In [6]:
data['class_size'].head()

CSD BOROUGH SCHOOL CODE                SCHOOL NAME GRADE  PROGRAM TYPE  \
0    1       M        M015  P.S. 015 Roberto Clemente     0K       GEN ED   
1    1       M        M015  P.S. 015 Roberto Clemente     0K          CTT   
2    1       M        M015  P.S. 015 Roberto Clemente     01       GEN ED   
3    1       M        M015  P.S. 015 Roberto Clemente     01          CTT   
4    1       M        M015  P.S. 015 Roberto Clemente     02       GEN ED   

  CORE SUBJECT (MS CORE and 9-12 ONLY) CORE COURSE (MS CORE and 9-12 ONLY)  \
0                                    -                                   -   
1                                    -                                   -   
2                                    -                                   -   
3                                    -                                   -   
4                                    -                                   -   

  SERVICE CATEGORY(K-9* ONLY)  NUMBER OF STUDENTS / SEATS FILLED  \
0                           -                               19.0   
1                           -                               21.0   
2                           -                               17.0   
3                           -                               17.0   
4                           -                               15.0   

   NUMBER OF SECTIONS  AVERAGE CLASS SIZE  SIZE OF SMALLEST CLASS  \
0                 1.0                19.0                    19.0   
1                 1.0                21.0                    21.0   
2                 1.0                17.0                    17.0   
3                 1.0                17.0                    17.0   
4                 1.0                15.0                    15.0   

   SIZE OF LARGEST CLASS DATA SOURCE  SCHOOLWIDE PUPIL-TEACHER RATIO  
0                   19.0         ATS                             NaN  
1                   21.0         ATS                             NaN  
2                   17.0         ATS                             NaN  
3                   17.0         ATS                             NaN  
4                   15.0         ATS                             NaN

Comparing these with the rows from the other datasets such as `sat_results`, we can infer that the `DBN` is just the combination of the `CSD` and `SCHOOL CODE` columns. However, the `DBN` values are padded, which means that the `CSD` portion contains two digits. If the `CSD` is less than two digits, we will have to add a leading 0. 

To do this, we will create a custom function that 
1. Takes in a number
2. Converts it to a string
3. Checks its length
4. Adds a leading 0 if the length is 1.

The leading 0 is added using the `str.zfill()` method. 

The padded numbers will then be combined with the `SCHOOL CODE` values to create a `DBN` column for `class_size`.

In [7]:
# Rename dbn column of `hs_directory`
data["hs_directory"].rename(columns={'dbn':'DBN'}, inplace=True) 

# Custom function that 'pads' the CSD numbers 
def pad_csd(num):
    string_num = str(num)
    if len(string_num) > 1:
        return string_num
    else:
        return string_num.zfill(2)
    
# Create a DBN column for class_size
data["class_size"]["padded_csd"] = data["class_size"]["CSD"].apply(pad_csd)
data["class_size"]["DBN"] = data["class_size"]["padded_csd"] + data["class_size"]["SCHOOL CODE"]

The first few rows of `class_size` are displayed to confirm the changes that we made. The new column should be at the end.

In [8]:
data["class_size"].head()

CSD BOROUGH SCHOOL CODE                SCHOOL NAME GRADE  PROGRAM TYPE  \
0    1       M        M015  P.S. 015 Roberto Clemente     0K       GEN ED   
1    1       M        M015  P.S. 015 Roberto Clemente     0K          CTT   
2    1       M        M015  P.S. 015 Roberto Clemente     01       GEN ED   
3    1       M        M015  P.S. 015 Roberto Clemente     01          CTT   
4    1       M        M015  P.S. 015 Roberto Clemente     02       GEN ED   

  CORE SUBJECT (MS CORE and 9-12 ONLY) CORE COURSE (MS CORE and 9-12 ONLY)  \
0                                    -                                   -   
1                                    -                                   -   
2                                    -                                   -   
3                                    -                                   -   
4                                    -                                   -   

  SERVICE CATEGORY(K-9* ONLY)  NUMBER OF STUDENTS / SEATS FILLED  \
0                           -                               19.0   
1                           -                               21.0   
2                           -                               17.0   
3                           -                               17.0   
4                           -                               15.0   

   NUMBER OF SECTIONS  AVERAGE CLASS SIZE  SIZE OF SMALLEST CLASS  \
0                 1.0                19.0                    19.0   
1                 1.0                21.0                    21.0   
2                 1.0                17.0                    17.0   
3                 1.0                17.0                    17.0   
4                 1.0                15.0                    15.0   

   SIZE OF LARGEST CLASS DATA SOURCE  SCHOOLWIDE PUPIL-TEACHER RATIO  \
0                   19.0         ATS                             NaN   
1                   21.0         ATS                             NaN   
2                   17.0         ATS                             NaN   
3                   17.0         ATS                             NaN   
4                   15.0         ATS                             NaN   

  padded_csd     DBN  
0         01  01M015  
1         01  01M015  
2         01  01M015  
3         01  01M015  
4         01  01M015

# Combining the SAT Scores
___

As previously mentioned, we will want to sum up the three SAT scores into a single one to make our analysis easier. Working with a single number will make it easier to find correlations with the factors specified in our project. 

We will first convert the values of the three columns to numeric types with the `pandas.to_numeric()` method so that we can add them. The `errors=coerce` argument is placed so that pandas will treat unconvertible strings as missing values. 

In [9]:
# Add the SAT scores to a single value after converting them to a numeric type
cols = ['SAT Math Avg. Score', 'SAT Critical Reading Avg. Score', 'SAT Writing Avg. Score']
for c in cols:
    data["sat_results"][c] = pd.to_numeric(data["sat_results"][c], errors="coerce")

data['sat_results']['sat_score'] = data['sat_results'][cols[0]] + data['sat_results'][cols[1]] + data['sat_results'][cols[2]]

The first few rows of `sat_results` are displayed to verify our changes. 

In [10]:
data['sat_results'].head()

DBN                                    SCHOOL NAME  \
0  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL STUDIES   
1  01M448            UNIVERSITY NEIGHBORHOOD HIGH SCHOOL   
2  01M450                     EAST SIDE COMMUNITY SCHOOL   
3  01M458                      FORSYTH SATELLITE ACADEMY   
4  01M509                        MARTA VALLE HIGH SCHOOL   

  Num of SAT Test Takers  SAT Critical Reading Avg. Score  \
0                     29                            355.0   
1                     91                            383.0   
2                     70                            377.0   
3                      7                            414.0   
4                     44                            390.0   

   SAT Math Avg. Score  SAT Writing Avg. Score  sat_score  
0                404.0                   363.0     1122.0  
1                423.0                   366.0     1172.0  
2                402.0                   370.0     1149.0  
3                401.0                   359.0     1174.0  
4                433.0                   384.0     1207.0

# Parsing Geographic Coordinates of Schools
___

We noticed that `hs_directory` has a column, `Location 1`, that contains the geographic coordinates of latitude and longitude for each school. The first few cells of this column are displayed below. 

In [11]:
pd.options.display.max_colwidth = 100 # To display all contents of each cell
data['hs_directory']['Location 1'].head()

0        883 Classon Avenue\nBrooklyn, NY 11225\n(40.67029890700047, -73.96164787599963)
1              1110 Boston Road\nBronx, NY 10456\n(40.8276026690005, -73.90447525699966)
2          1501 Jerome Avenue\nBronx, NY 10452\n(40.842414068000494, -73.91616158599965)
3          411 Pearl Street\nNew York, NY 10038\n(40.71067947100045, -74.00080702099967)
4    160-20 Goethals Avenue\nJamaica, NY 11432\n(40.718810094000446, -73.80650045499965)
Name: Location 1, dtype: object

The coordinates are located within parentheses at the end of each string. We will extract these values to map the schools which will then help us discover any geographic patterns within the data. 

To do this, we can use the vectorized string method `str.extract()` to extract each coordinate. Both will make use of the regular expression: `\(.+, .+\)` to isolate the coordinates portion of each cell in `Location 1`. Since parentheses have a special meaning in regular expressions, the backslashes cancel or escape this meaning and allows for the parentheses to be taken literally. The two `.+`s represent the latitude and longitude values which are separated by a comma, `,`. 

Capture groups (parentheses where each is NOT preceded by a backslash) will allow us to take either the latitude or the longitude, separately.  

In [12]:
# Extract the latitudes and longitudes separately and store them in two new columns
data["hs_directory"]["lat"] = data["hs_directory"]["Location 1"].str.extract(r"\((.+), .+\)")[0] 
data["hs_directory"]["lon"] = data["hs_directory"]["Location 1"].str.extract(r"\(.+, (.+)\)")[0]

print(data["hs_directory"]["lat"].dtype)
print(data["hs_directory"]["lon"].dtype)

object
object


The `[0]` portion at the end of each line above is needed as the `str.extract()` method returns a DataFrame with a column named `0`. 

Since the `lat` and `lon` columns are of the type `object`, they are converted to numeric types below. 

The first few values for each are displayed to confirm our extraction process. 

In [13]:
data["hs_directory"]["lat"] = pd.to_numeric(data["hs_directory"]["lat"], errors="coerce")
data["hs_directory"]["lon"] = pd.to_numeric(data["hs_directory"]["lon"], errors="coerce")

data["hs_directory"][['lat','lon']].head()

lat        lon
0  40.670299 -73.961648
1  40.827603 -73.904475
2  40.842414 -73.916162
3  40.710679 -74.000807
4  40.718810 -73.806500

# Condense datasets
___

Since there are data sets with high schools that appear multiple times, this behavior carries over to the DBN values as well. Combining data sets without addressing this issue will result into problems. We would be merging a single row in one data set with two or more rows in another data set, which is not possible. 

Hence, we will have to condense the data sets with this multiplicity issue. Namely, these are:
- `class_size`
- `graduation`
- `demographics`

For `class_size`, the first few rows are displayed below.

In [14]:
data['class_size'].head()

CSD BOROUGH SCHOOL CODE                SCHOOL NAME GRADE  PROGRAM TYPE  \
0    1       M        M015  P.S. 015 Roberto Clemente     0K       GEN ED   
1    1       M        M015  P.S. 015 Roberto Clemente     0K          CTT   
2    1       M        M015  P.S. 015 Roberto Clemente     01       GEN ED   
3    1       M        M015  P.S. 015 Roberto Clemente     01          CTT   
4    1       M        M015  P.S. 015 Roberto Clemente     02       GEN ED   

  CORE SUBJECT (MS CORE and 9-12 ONLY) CORE COURSE (MS CORE and 9-12 ONLY)  \
0                                    -                                   -   
1                                    -                                   -   
2                                    -                                   -   
3                                    -                                   -   
4                                    -                                   -   

  SERVICE CATEGORY(K-9* ONLY)  NUMBER OF STUDENTS / SEATS FILLED  \
0                           -                               19.0   
1                           -                               21.0   
2                           -                               17.0   
3                           -                               17.0   
4                           -                               15.0   

   NUMBER OF SECTIONS  AVERAGE CLASS SIZE  SIZE OF SMALLEST CLASS  \
0                 1.0                19.0                    19.0   
1                 1.0                21.0                    21.0   
2                 1.0                17.0                    17.0   
3                 1.0                17.0                    17.0   
4                 1.0                15.0                    15.0   

   SIZE OF LARGEST CLASS DATA SOURCE  SCHOOLWIDE PUPIL-TEACHER RATIO  \
0                   19.0         ATS                             NaN   
1                   21.0         ATS                             NaN   
2                   17.0         ATS                             NaN   
3                   17.0         ATS                             NaN   
4                   15.0         ATS                             NaN   

  padded_csd     DBN  
0         01  01M015  
1         01  01M015  
2         01  01M015  
3         01  01M015  
4         01  01M015

Unlike in `sat_results`, this data set contains non-high school information as indicated by the grade levels below. 

In [15]:
data['class_size']['GRADE '].unique() # Take note of the space after GRADE

array(['0K', '01', '02', '03', '04', '05', '0K-09', nan, '06', '07', '08',
       'MS Core', '09-12', '09'], dtype=object)

There are also different types of programs offered by each school. 

In [16]:
data['class_size']['PROGRAM TYPE'].value_counts()

GEN ED     14545
CTT         7460
SPEC ED     3653
G&T          469
Name: PROGRAM TYPE, dtype: int64

Since we are interested in high schools, we will select those rows with `09-12` for the `GRADE ` column. We will also choose `GEN ED` under `PROGRAM TYPE` as a second filter since the others are special types of educational programs. 

In [17]:
# Filter class_size by GRADE and PROGRAM TYPE
class_size = data["class_size"]
class_size = class_size[class_size["GRADE "] == "09-12"]
class_size = class_size[class_size["PROGRAM TYPE"] == "GEN ED"]
class_size.head()

CSD BOROUGH SCHOOL CODE                                    SCHOOL NAME  \
225    1       M        M292  Henry Street School for International Studies   
226    1       M        M292  Henry Street School for International Studies   
227    1       M        M292  Henry Street School for International Studies   
228    1       M        M292  Henry Street School for International Studies   
229    1       M        M292  Henry Street School for International Studies   

    GRADE  PROGRAM TYPE CORE SUBJECT (MS CORE and 9-12 ONLY)  \
225  09-12       GEN ED                              ENGLISH   
226  09-12       GEN ED                              ENGLISH   
227  09-12       GEN ED                              ENGLISH   
228  09-12       GEN ED                              ENGLISH   
229  09-12       GEN ED                                 MATH   

    CORE COURSE (MS CORE and 9-12 ONLY) SERVICE CATEGORY(K-9* ONLY)  \
225                           English 9                           -   
226                          English 10                           -   
227                          English 11                           -   
228                          English 12                           -   
229                  Integrated Algebra                           -   

     NUMBER OF STUDENTS / SEATS FILLED  NUMBER OF SECTIONS  \
225                               63.0                 3.0   
226                               79.0                 3.0   
227                               38.0                 2.0   
228                               69.0                 3.0   
229                               53.0                 3.0   

     AVERAGE CLASS SIZE  SIZE OF SMALLEST CLASS  SIZE OF LARGEST CLASS  \
225                21.0                    19.0                   25.0   
226                26.3                    24.0                   31.0   
227                19.0                    16.0                   22.0   
228                23.0                    13.0                   30.0   
229                17.7                    16.0                   21.0   

    DATA SOURCE  SCHOOLWIDE PUPIL-TEACHER RATIO padded_csd     DBN  
225       STARS                             NaN         01  01M292  
226       STARS                             NaN         01  01M292  
227       STARS                             NaN         01  01M292  
228       STARS                             NaN         01  01M292  
229       STARS                             NaN         01  01M292

Above, we can see that there are still duplicate DBN values. This is due to the other columns such as `CORE SUBJECT (MS CORE and 9-12 ONLY)` and `CORE COURSE (MS CORE and 9-12 ONLY)`. It appears that these columns refer to different subjects and courses offered by each school. 

We want to have a single occurrence for each DBN value while somehow retaining the information contained in the data set above. Since we are more concerned of the numeric columns (that pertain to class size), we can condense the filtered data set by finding the averages of these columns. 

To do this, we will make use of the `pandas.DataFrame.groupby()` method. We will use this to group the data set by DBN value. We will then chain the `agg()` method, passing the `numpy.mean()`function as an argument (without parentheses), to compute the averages per DBN.  

In [18]:
# Average the numeric columns to per DBN value
class_size = class_size.groupby("DBN").agg(np.mean)
class_size.reset_index(inplace=True)

data["class_size"] = class_size
data["class_size"].head()

DBN  CSD  NUMBER OF STUDENTS / SEATS FILLED  NUMBER OF SECTIONS  \
0  01M292    1                            88.0000            4.000000   
1  01M332    1                            46.0000            2.000000   
2  01M378    1                            33.0000            1.000000   
3  01M448    1                           105.6875            4.750000   
4  01M450    1                            57.6000            2.733333   

   AVERAGE CLASS SIZE  SIZE OF SMALLEST CLASS  SIZE OF LARGEST CLASS  \
0           22.564286                   18.50              26.571429   
1           22.000000                   21.00              23.500000   
2           33.000000                   33.00              33.000000   
3           22.231250                   18.25              27.062500   
4           21.200000                   19.40              22.866667   

   SCHOOLWIDE PUPIL-TEACHER RATIO  
0                             NaN  
1                             NaN  
2                             NaN  
3                             NaN  
4                             NaN

We now have single occurrences of the DBN values for `class_size`. Note that the non-numeric columns are not included in the resulting dataset. 

Moving on to `demographics`, the first few rows are displayed below.

In [19]:
data["demographics"].head(7)

DBN                                              Name  schoolyear  \
0  01M015                         P.S. 015 ROBERTO CLEMENTE    20052006   
1  01M015                         P.S. 015 ROBERTO CLEMENTE    20062007   
2  01M015                         P.S. 015 ROBERTO CLEMENTE    20072008   
3  01M015                         P.S. 015 ROBERTO CLEMENTE    20082009   
4  01M015                         P.S. 015 ROBERTO CLEMENTE    20092010   
5  01M015                         P.S. 015 ROBERTO CLEMENTE    20102011   
6  01M015  P.S. 015 ROBERTO CLEMENTE                           20112012   

  fl_percent  frl_percent  total_enrollment prek   k grade1 grade2 grade3  \
0       89.4          NaN               281   15  36     40     33     38   
1       89.4          NaN               243   15  29     39     38     34   
2       89.4          NaN               261   18  43     39     36     38   
3       89.4          NaN               252   17  37     44     32     34   
4                    96.5               208   16  40     28     32     30   
5                    96.5               203   13  37     35     33     30   
6        NaN         89.4               189   13  31     35     28     25   

  grade4 grade5 grade6 grade7 grade8 grade9 grade10 grade11 grade12  ell_num  \
0     52     29     38    NaN    NaN    NaN     NaN     NaN     NaN     36.0   
1     42     46    NaN    NaN    NaN    NaN     NaN     NaN     NaN     38.0   
2     47     40    NaN    NaN    NaN    NaN     NaN     NaN     NaN     52.0   
3     39     49    NaN    NaN    NaN    NaN     NaN     NaN     NaN     48.0   
4     24     38    NaN    NaN    NaN    NaN     NaN     NaN     NaN     40.0   
5     30     25    NaN    NaN    NaN    NaN     NaN     NaN     NaN     30.0   
6     28     29                                                         20.0   

   ell_percent  sped_num  sped_percent ctt_num selfcontained_num  asian_num  \
0         12.8      57.0          20.3      25                 9         10   
1         15.6      55.0          22.6      19                15         18   
2         19.9      60.0          23.0      20                14         16   
3         19.0      62.0          24.6      21                17         16   
4         19.2      46.0          22.1      14                14         16   
5         14.8      46.0          22.7      21                 9         13   
6         10.6      40.0          21.2      23                 7         12   

   asian_per  black_num  black_per  hispanic_num  hispanic_per  white_num  \
0        3.6         74       26.3           189          67.3          5   
1        7.4         68       28.0           153          63.0          4   
2        6.1         77       29.5           157          60.2          7   
3        6.3         75       29.8           149          59.1          7   
4        7.7         67       32.2           118          56.7          6   
5        6.4         75       36.9           110          54.2          4   
6        6.3         63       33.3           109          57.7          4   

   white_per  male_num  male_per  female_num  female_per  
0        1.8     158.0      56.2       123.0        43.8  
1        1.6     140.0      57.6       103.0        42.4  
2        2.7     143.0      54.8       118.0        45.2  
3        2.8     149.0      59.1       103.0        40.9  
4        2.9     124.0      59.6        84.0        40.4  
5        2.0     113.0      55.7        90.0        44.3  
6        2.1      97.0      51.3        92.0        48.7

We can see that there are multiple values for `schoolyear` per school. We will select the most recent school year, `20112012` which will also match our SAT results data set (2012). 

In [20]:
data["demographics"] = data["demographics"][data["demographics"]["schoolyear"] == 20112012]
data["demographics"].head()

DBN                                              Name  schoolyear  \
6   01M015  P.S. 015 ROBERTO CLEMENTE                           20112012   
13  01M019  P.S. 019 ASHER LEVY                                 20112012   
20  01M020  PS 020 ANNA SILVER                                  20112012   
27  01M034  PS 034 FRANKLIN D ROOSEVELT                         20112012   
35  01M063  PS 063 WILLIAM MCKINLEY                             20112012   

   fl_percent  frl_percent  total_enrollment prek    k grade1 grade2 grade3  \
6         NaN         89.4               189   13   31     35     28     25   
13        NaN         61.5               328   32   46     52     54     52   
20        NaN         92.5               626   52  102    121     87     88   
27        NaN         99.7               401   14   34     38     36     45   
35        NaN         78.9               176   18   20     30     21     31   

   grade4 grade5 grade6 grade7 grade8 grade9 grade10 grade11 grade12  ell_num  \
6      28     29                                                         20.0   
13     46     46                                                         33.0   
20     85     91                                                        128.0   
27     28     40     55     55     56                                    34.0   
35     26     30                                                          6.0   

    ell_percent  sped_num  sped_percent ctt_num selfcontained_num  asian_num  \
6          10.6      40.0          21.2      23                 7         12   
13         10.1      59.0          18.0      16                16         51   
20         20.4      97.0          15.5      49                31        190   
27          8.5     106.0          26.4      59                16         22   
35          3.4      45.0          25.6      34                 4          9   

    asian_per  black_num  black_per  hispanic_num  hispanic_per  white_num  \
6         6.3         63       33.3           109          57.7          4   
13       15.5         81       24.7           158          48.2         28   
20       30.4         55        8.8           357          57.0         16   
27        5.5         90       22.4           275          68.6          8   
35        5.1         41       23.3           110          62.5         15   

    white_per  male_num  male_per  female_num  female_per  
6         2.1      97.0      51.3        92.0        48.7  
13        8.5     147.0      44.8       181.0        55.2  
20        2.6     330.0      52.7       296.0        47.3  
27        2.0     204.0      50.9       197.0        49.1  
35        8.5      97.0      55.1        79.0        44.9

Finally, we move on to `graduation`. 

In [21]:
data["graduation"].head()

Demographic     DBN                            School Name    Cohort  \
0  Total Cohort  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL      2003   
1  Total Cohort  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL      2004   
2  Total Cohort  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL      2005   
3  Total Cohort  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL      2006   
4  Total Cohort  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL  2006 Aug   

   Total Cohort Total Grads - n Total Grads - % of cohort Total Regents - n  \
0             5               s                         s                 s   
1            55              37                     67.3%                17   
2            64              43                     67.2%                27   
3            78              43                     55.1%                36   
4            78              44                     56.4%                37   

  Total Regents - % of cohort Total Regents - % of grads Advanced Regents - n  \
0                           s                          s                    s   
1                       30.9%                      45.9%                    0   
2                       42.2%                      62.8%                    0   
3                       46.2%                      83.7%                    0   
4                       47.4%                      84.1%                    0   

  Advanced Regents - % of cohort Advanced Regents - % of grads  \
0                              s                             s   
1                             0%                            0%   
2                             0%                            0%   
3                             0%                            0%   
4                             0%                            0%   

  Regents w/o Advanced - n Regents w/o Advanced - % of cohort  \
0                        s                                  s   
1                       17                              30.9%   
2                       27                              42.2%   
3                       36                              46.2%   
4                       37                              47.4%   

  Regents w/o Advanced - % of grads Local - n Local - % of cohort  \
0                                 s         s                   s   
1                             45.9%        20               36.4%   
2                             62.8%        16                 25%   
3                             83.7%         7                  9%   
4                             84.1%         7                  9%   

    Local - % of grads Still Enrolled - n Still Enrolled - % of cohort  \
0                    s                  s                            s   
1                54.1%                 15                        27.3%   
2  37.200000000000003%                  9                        14.1%   
3                16.3%                 16                        20.5%   
4                15.9%                 15                        19.2%   

  Dropped Out - n Dropped Out - % of cohort  
0               s                         s  
1               3                      5.5%  
2               9                     14.1%  
3              11                     14.1%  
4              11                     14.1%

We are concerned with the `Demographic` and `Cohort` columns, which refer to a specific demographic group and to the year of the data, respectively. 

In [22]:
data["graduation"].Demographic.unique()

array(['Total Cohort', 'Male', 'English Language Learners',
       'English Proficient Students', 'Special Education Students',
       'General Education Students', 'Asian', 'Black', 'Hispanic',
       'White', 'Female'], dtype=object)

In [23]:
data["graduation"].Cohort.unique()

array(['2003', '2004', '2005', '2006', '2006 Aug', '2001', '2002'],
      dtype=object)

Just like in `demographics`, we will want to select the latest year, which happens to be 2006. We will also select `Total Cohort` for the `Demographic` column to include all the students. 

In [24]:
# Filter the graduation data set
data["graduation"] = data["graduation"][data["graduation"]["Cohort"] == "2006"]
data["graduation"] = data["graduation"][data["graduation"]["Demographic"] == "Total Cohort"]

data["graduation"].head()

Demographic     DBN                            School Name Cohort  \
3   Total Cohort  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL   2006   
10  Total Cohort  01M448    UNIVERSITY NEIGHBORHOOD HIGH SCHOOL   2006   
17  Total Cohort  01M450             EAST SIDE COMMUNITY SCHOOL   2006   
24  Total Cohort  01M509                MARTA VALLE HIGH SCHOOL   2006   
31  Total Cohort  01M515  LOWER EAST SIDE PREPARATORY HIGH SCHO   2006   

    Total Cohort Total Grads - n Total Grads - % of cohort Total Regents - n  \
3             78              43                     55.1%                36   
10           124              53                     42.7%                42   
17            90              70                     77.8%                67   
24            84              47                       56%                40   
31           193             105                     54.4%                91   

   Total Regents - % of cohort Total Regents - % of grads  \
3                        46.2%                      83.7%   
10                       33.9%                      79.2%   
17         74.400000000000006%                      95.7%   
24                       47.6%                      85.1%   
31                       47.2%                      86.7%   

   Advanced Regents - n Advanced Regents - % of cohort  \
3                     0                             0%   
10                    8                           6.5%   
17                    0                             0%   
24                   17                          20.2%   
31                   69            35.799999999999997%   

   Advanced Regents - % of grads Regents w/o Advanced - n  \
3                             0%                       36   
10                         15.1%                       34   
17                            0%                       67   
24           36.200000000000003%                       23   
31                         65.7%                       22   

   Regents w/o Advanced - % of cohort Regents w/o Advanced - % of grads  \
3                               46.2%                             83.7%   
10                              27.4%                             64.2%   
17                74.400000000000006%                             95.7%   
24                              27.4%                             48.9%   
31                              11.4%                               21%   

   Local - n Local - % of cohort Local - % of grads Still Enrolled - n  \
3          7                  9%              16.3%                 16   
10        11                8.9%              20.8%                 46   
17         3                3.3%               4.3%                 15   
24         7  8.300000000000001%              14.9%                 25   
31        14                7.3%              13.3%                 53   

   Still Enrolled - % of cohort Dropped Out - n Dropped Out - % of cohort  
3                         20.5%              11                     14.1%  
10                        37.1%              20       16.100000000000001%  
17                        16.7%               5                      5.6%  
24                        29.8%               5                        6%  
31                        27.5%              35       18.100000000000001%

We have now successfully condensed the data sets of `class_size`, `demographics`, and `graduation`. 

# Convert AP scores to numeric
___

The last data set we will clean before combining everything is `ap_2010`. Just as in `sat_results`, we will convert some columns for this data set into the numeric type. On a scale of 1-5, 3 is the passing score for AP exams. 

This [website](https://www.ivywise.com/blog/test-prep-101-ap-exams-vs-sat-subject-tests/) contains a table that presents the differences between AP exams and the SAT.

In [25]:
data['ap_2010'].head()

DBN                             SchoolName AP Test Takers   \
0  01M448           UNIVERSITY NEIGHBORHOOD H.S.              39   
1  01M450                 EAST SIDE COMMUNITY HS              19   
2  01M515                    LOWER EASTSIDE PREP              24   
3  01M539         NEW EXPLORATIONS SCI,TECH,MATH             255   
4  02M296  High School of Hospitality Management               s   

  Total Exams Taken Number of Exams with scores 3 4 or 5  
0                49                                   10  
1                21                                    s  
2                26                                   24  
3               377                                  191  
4                 s                                    s

We convert the pertinent columns to the numeric type below. 

In [26]:
cols = ['AP Test Takers ', 'Total Exams Taken', 'Number of Exams with scores 3 4 or 5']

for col in cols:
    data["ap_2010"][col] = pd.to_numeric(data["ap_2010"][col], errors="coerce")
    
data["ap_2010"][cols].dtypes

AP Test Takers                          float64
Total Exams Taken                       float64
Number of Exams with scores 3 4 or 5    float64
dtype: object

# Combine the data sets
___

The `DBN` column will serve as the reference for our combining process. This will allow us to combine one row in a data set with another row in another data set if both contain the same DBN value. Do note, though, that DBN values may be present in one data set but absent in another. This is due to factors such as human error in data gathering, etc. 

We will use the `pandas.DataFrame.merge()` function to join our data sets. We need to be wary of which join type to use from `left`, `right`, `inner`, or `outer` as each one behaves differently and has its own pros and cons. Considering this behavior and the inconsistency of DBN values across the data sets, we cannot use the same join type for all data sets. Hence, we will combine the data sets gradually, using the appropriate join type as we go along. 

We start by creating a variable, `combined`, to contain our final data set. `sat_results` is placed there first as this is our primary data set. 

In [27]:
combined = data["sat_results"]

We will perform the `left` joins first. This type of join allows us to retain all the rows of `sat_results`. This is most appropriate when combining our primary data set with those that have missing or fewer DBN values. Although this may cause us to lose rows from the other data set if there are DBN values there not found in `sat_results`, this is better than losing data from our primary data set. 

Below, we see that `ap_2010` and `graduation` fall under this criteria. 

In [28]:
# Print the number of DBN values per data set
data_sets = ['sat_results', 'ap_2010', 'graduation']

for d in data_sets:
    print('{0}{1} {2}'.format(d,':', data[d].DBN.count()))

sat_results: 478
ap_2010: 258
graduation: 405


We now do the `left` merge process below. 

In [29]:
combined = combined.merge(data["ap_2010"], on="DBN", how="left")
combined = combined.merge(data["graduation"], on="DBN", how="left")

combined.head()

DBN                                    SCHOOL NAME  \
0  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL STUDIES   
1  01M448            UNIVERSITY NEIGHBORHOOD HIGH SCHOOL   
2  01M450                     EAST SIDE COMMUNITY SCHOOL   
3  01M458                      FORSYTH SATELLITE ACADEMY   
4  01M509                        MARTA VALLE HIGH SCHOOL   

  Num of SAT Test Takers  SAT Critical Reading Avg. Score  \
0                     29                            355.0   
1                     91                            383.0   
2                     70                            377.0   
3                      7                            414.0   
4                     44                            390.0   

   SAT Math Avg. Score  SAT Writing Avg. Score  sat_score  \
0                404.0                   363.0     1122.0   
1                423.0                   366.0     1172.0   
2                402.0                   370.0     1149.0   
3                401.0                   359.0     1174.0   
4                433.0                   384.0     1207.0   

                     SchoolName  AP Test Takers   Total Exams Taken  \
0                           NaN              NaN                NaN   
1  UNIVERSITY NEIGHBORHOOD H.S.             39.0               49.0   
2        EAST SIDE COMMUNITY HS             19.0               21.0   
3                           NaN              NaN                NaN   
4                           NaN              NaN                NaN   

   Number of Exams with scores 3 4 or 5   Demographic  \
0                                   NaN  Total Cohort   
1                                  10.0  Total Cohort   
2                                   NaN  Total Cohort   
3                                   NaN           NaN   
4                                   NaN  Total Cohort   

                             School Name Cohort  Total Cohort Total Grads - n  \
0  HENRY STREET SCHOOL FOR INTERNATIONAL   2006          78.0              43   
1    UNIVERSITY NEIGHBORHOOD HIGH SCHOOL   2006         124.0              53   
2             EAST SIDE COMMUNITY SCHOOL   2006          90.0              70   
3                                    NaN    NaN           NaN             NaN   
4                MARTA VALLE HIGH SCHOOL   2006          84.0              47   

  Total Grads - % of cohort Total Regents - n Total Regents - % of cohort  \
0                     55.1%                36                       46.2%   
1                     42.7%                42                       33.9%   
2                     77.8%                67         74.400000000000006%   
3                       NaN               NaN                         NaN   
4                       56%                40                       47.6%   

  Total Regents - % of grads Advanced Regents - n  \
0                      83.7%                    0   
1                      79.2%                    8   
2                      95.7%                    0   
3                        NaN                  NaN   
4                      85.1%                   17   

  Advanced Regents - % of cohort Advanced Regents - % of grads  \
0                             0%                            0%   
1                           6.5%                         15.1%   
2                             0%                            0%   
3                            NaN                           NaN   
4                          20.2%           36.200000000000003%   

  Regents w/o Advanced - n Regents w/o Advanced - % of cohort  \
0                       36                              46.2%   
1                       34                              27.4%   
2                       67                74.400000000000006%   
3                      NaN                                NaN   
4                       23                              27.4%   

  Regents w/o Advanced - % of grads Local - n Local - % of cohort  \
0              

In [30]:
combined.shape

(479, 33)

For the other data sets, we can see that there are more DBN values. These data sets also hold more valuable information for our analysis. 

In [31]:
# Print the number of DBN values per data set
to_merge = ["class_size", "demographics", "survey", "hs_directory"]

for d in to_merge:
    print('{0}{1} {2}'.format(d,':', data[d].DBN.count()))

class_size: 583
demographics: 1509
survey: 1646
hs_directory: 435


To retain as much information, we will use the `inner` join for these data sets. This join type allows us to retain rows that have DBN values common to the two data sets being combined at a given merging process. 

The final data set is displayed after to confirm the changes we made. 

In [32]:
to_merge = ["class_size", "demographics", "survey", "hs_directory"]

for m in to_merge:
    combined = combined.merge(data[m], on="DBN", how="inner")
    
combined.head()

DBN                                                     SCHOOL NAME  \
0  01M292                   HENRY STREET SCHOOL FOR INTERNATIONAL STUDIES   
1  01M448                             UNIVERSITY NEIGHBORHOOD HIGH SCHOOL   
2  01M450                                      EAST SIDE COMMUNITY SCHOOL   
3  01M509                                         MARTA VALLE HIGH SCHOOL   
4  01M539  NEW EXPLORATIONS INTO SCIENCE, TECHNOLOGY AND MATH HIGH SCHOOL   

  Num of SAT Test Takers  SAT Critical Reading Avg. Score  \
0                     29                            355.0   
1                     91                            383.0   
2                     70                            377.0   
3                     44                            390.0   
4                    159                            522.0   

   SAT Math Avg. Score  SAT Writing Avg. Score  sat_score  \
0                404.0                   363.0     1122.0   
1                423.0                   366.0     1172.0   
2                402.0                   370.0     1149.0   
3                433.0                   384.0     1207.0   
4                574.0                   525.0     1621.0   

                       SchoolName  AP Test Takers   Total Exams Taken  \
0                             NaN              NaN                NaN   
1    UNIVERSITY NEIGHBORHOOD H.S.             39.0               49.0   
2          EAST SIDE COMMUNITY HS             19.0               21.0   
3                             NaN              NaN                NaN   
4  NEW EXPLORATIONS SCI,TECH,MATH            255.0              377.0   

   Number of Exams with scores 3 4 or 5   Demographic  \
0                                   NaN  Total Cohort   
1                                  10.0  Total Cohort   
2                                   NaN  Total Cohort   
3                                   NaN  Total Cohort   
4                                 191.0  Total Cohort   

                             School Name Cohort  Total Cohort Total Grads - n  \
0  HENRY STREET SCHOOL FOR INTERNATIONAL   2006          78.0              43   
1    UNIVERSITY NEIGHBORHOOD HIGH SCHOOL   2006         124.0              53   
2             EAST SIDE COMMUNITY SCHOOL   2006          90.0              70   
3                MARTA VALLE HIGH SCHOOL   2006          84.0              47   
4  NEW EXPLORATIONS INTO SCIENCE  TECHNO   2006          46.0              46   

  Total Grads - % of cohort Total Regents - n Total Regents - % of cohort  \
0                     55.1%                36                       46.2%   
1                     42.7%                42                       33.9%   
2                     77.8%                67         74.400000000000006%   
3                       56%                40                       47.6%   
4                      100%                46                        100%   

  Total Regents - % of grads Advanced Regents - n  \
0                      83.7%                    0   
1                      79.2%                    8   
2                      95.7%                    0   
3                      85.1%                   17   
4                       100%                   31   

  Advanced Regents - % of cohort Advanced Regents - % of grads  \
0                             0%                            0%   
1                           6.5%                         15.1%   
2                             0%                            0%   
3                          20.2%           36.200000000000003%   
4            67.400000000000006%           67.400000000000006%   

  Regents w/o Advanced - n Regents w/o Advanced - % of cohort  \
0                       36                              46.2%   
1                       34                              27.4%   
2                       67                74.400000000000006%   
3                       23                              27.4%   
4                       15          

In [33]:
combined.shape

(363, 158)

We can notice that, after the `inner` join, the number of rows has decreased from the initial 479 to 363. This is because of the DBN values in `sat_results` that could not be found in the other data sets. This is fine for our purposes since we are interested in high-level correlations.

There are also several missing values in the data set. Some columns are entirely missing values. This is due to the previous cleaning and merging steps we did.

Pandas will not be able to find correlations if there are missing values. To handle this, we will replace `NaN` with the mean of the column, wherever possible. The rest will be replaced with 0.     

In [34]:
# Replace missing values with the mean and with 0
combined = combined.fillna(combined.mean())
combined = combined.fillna(0)

combined.notnull().describe()

DBN SCHOOL NAME Num of SAT Test Takers  \
count    363         363                    363   
unique     1           1                      1   
top     True        True                   True   
freq     363         363                    363   

       SAT Critical Reading Avg. Score SAT Math Avg. Score  \
count                              363                 363   
unique                               1                   1   
top                               True                True   
freq                               363                 363   

       SAT Writing Avg. Score sat_score SchoolName AP Test Takers   \
count                     363       363        363             363   
unique                      1         1          1               1   
top                      True      True       True            True   
freq                      363       363        363             363   

       Total Exams Taken Number of Exams with scores 3 4 or 5 Demographic  \
count                363                                  363         363   
unique                 1                                    1           1   
top                 True                                 True        True   
freq                 363                                  363         363   

       School Name Cohort Total Cohort Total Grads - n  \
count          363    363          363             363   
unique           1      1            1               1   
top           True   True         True            True   
freq           363    363          363             363   

       Total Grads - % of cohort Total Regents - n  \
count                        363               363   
unique                         1                 1   
top                         True              True   
freq                         363               363   

       Total Regents - % of cohort Total Regents - % of grads  \
count                          363                        363   
unique                           1                          1   
top                           True                       True   
freq                           363                        363   

       Advanced Regents - n Advanced Regents - % of cohort  \
count                   363                            363   
unique                    1                              1   
top                    True                           True   
freq                    363                            363   

       Advanced Regents - % of grads Regents w/o Advanced - n  \
count                            363                      363   
unique                             1                        1   
top                             True                     True   
freq                             363                      363   

       Regents w/o Advanced - % of cohort Regents w/o Advanced - % of grads  \
count                                 363                               363   
unique                                  1                                 1   
top                                  True                              True   
freq                                  363                               363   

       Local - n Local - % of cohort Local - % of grads Still Enrolled - n  \
count        363                 363                363                363   
unique         1                   1                  1                  1   
top         True                True               True               True   
freq         363                 363                363                363   

       Still Enrolled - % of cohort Dropped Out - n Dropped Out - % of cohort  \
count                           363             363                       363   
unique                            1               1                         1   
top                            True            True                      True   
freq                            363             363                  

We can confirm from the output above that all missing values have now been replaced. 

# Create a school district column for mapping
___

A good way to map the schools would be on a district level. A school's district is simply the first two digits of the DBN, or the CSD. We can simply rename the `CSD` column to `school_dist` to indicate this. 

In [37]:
combined.rename(columns = {'CSD':'school_dist'}, inplace = True)

combined[["DBN","school_dist"]].head(10)

DBN  school_dist
0  01M292            1
1  01M448            1
2  01M450            1
3  01M509            1
4  01M539            1
5  01M696            1
6  02M047            2
7  02M288            2
8  02M294            2
9  02M296            2

The final dataset is converted below to a `csv` file to be used for [analysis](Data_Analysis.ipynb).

In [38]:
combined.to_csv('combined.csv', index = False)